Added basic CRUD operations

Htrap19 opened the pull request:
Added basic CRUD operations + relations for Project Entity

Htrap19 committed:
Basic CRUD operations added

Mknan-fms commented on pom.xml:
```diff
@@ -19,7 +19,23 @@
+		<dependency>
+			<groupId>com.h2database</groupId>
+			<artifactId>h2</artifactId>
+			<version>2.0.204</version>
```
Let the framework decide which version of the DB is suitable. 
It should be provided on **RUNTIME** scope.
Also, we can import **DEVTOOLS-STARTER** dependency, since it adds the feature of live-reload after code change

Htrap19 replied:
Sure

Mknan-fms commented on src/main/java/com/interfaceict/ptrakcer/controllers/ProjectEntityController.java:
```diff
@@ -0,0 +1,32 @@
+    private ProjectEntityServiceImpl m_Service;
+
+    @GetMapping("/projects")
+    public List<ProjectEntity> getAllProjects() { return m_Service.getAll(); }
```
For sample projects we can just return original entity, while for real projects it is a good-practice to use a DTO as return type

Htrap19 replied:
Yep i thought of using DTO but for now DTO will be useless

Mknan-fms replied:
When entities numbers grow, we have to adapt DTO.

Mknan-fms commented on src/main/java/com/interfaceict/ptrakcer/enums/ApplicationType.java:
```diff
@@ -0,0 +1,14 @@
+
+package com.interfaceict.ptrakcer.enums;
+
+public enum ApplicationType
```
ENUM values are better to be upper-case

Mknan-fms commented on src/main/java/com/interfaceict/ptrakcer/controllers/ProjectEntityController.java:
```diff
@@ -0,0 +1,32 @@
+    public List<ProjectEntity> getAllProjects() { return m_Service.getAll(); }
+
+    @PostMapping("/projects")
+    public void newProject(@Valid @RequestBody ProjectEntity newProject) { m_Service.save(newProject); }
```
You could use a simple **POJO** instead of the real entity **POJO**

Htrap19 replied:
Also can we use DTO ?

Mknan-fms commented on src/main/java/com/interfaceict/ptrakcer/models/ProjectSoftwareDetails.java:
```diff
@@ -0,0 +1,93 @@
+    @Column(name = "framework")
+    private String m_FrameworkUsed;
+
+    // == Database details == - in future it might have a separate model
```
If it's better to have a different model, then do it immediately!

Mknan-fms commented on src/main/java/com/interfaceict/ptrakcer/repositories/ProjectEntityRepo.java:
```diff
@@ -0,0 +1,12 @@
+
+import com.interfaceict.ptrakcer.models.ProjectEntity;
+import org.springframework.data.jpa.repository.JpaRepository;
+
```
It should have an `@Repository` annotation

Mknan-fms commented on src/main/java/com/interfaceict/ptrakcer/services/ProjectService.java:
```diff
@@ -0,0 +1,23 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class ProjectService<T, Repo extends JpaRepository<T, Long>>
```
Some duplication here, `Repo extends JpaRepository<T , Long>` also `ProjectEntityRepo` do the same functionality. It's better to use one of them.

Mknan-fms commented on src/main/java/com/interfaceict/ptrakcer/services/utils/Utils.java:
```diff
@@ -0,0 +1,19 @@
+
+public class Utils
+{
+    static public <T> Boolean CheckIfItsNullOrEmpty(T obj)
```
There are multiple validation rules provided by the `spring-boot-starter-validation` dependency. 
For example, it's possible to replace `CheckIfItsNullOrEmpty` validation function with just two annotation:
`@NotEmpty` or `@NotNull`

Mknan-fms commented on src/main/java/com/interfaceict/ptrakcer/servicesimpl/ProjectEntityServiceImpl.java:
```diff
@@ -0,0 +1,50 @@
+
+        ProjectEntity entityDB = entityDBOpt.get();
+
+        // This can be refactored more
```
Validation rules varies from numeric types to date type to string types

Mknan-fms reviewed (commented):
- `DTOs` are better to be used for API request/response objects
- Use framework-provided validation annotations instead of developing custom validation rules, unless there's a true need for providing such custom validation logic
- Follow framework guidelines when working with `JPA Repositories`

Htrap19 committed:
Minor fixes + added relations for Project entity

Mknan-fms committed:
Minor changes

Mknan-fms committed:
Merge branch 'crud' of https://github.com/Mknan-fms/project-tracker into crud

Mknan-fms committed:
Update .gitignore

Mknan-fms reviewed (changes requested):
- How related entities are added? { `Assests` , `Dependency` , ...etc } missing endpoints for achieving such actions

Mknan-fms committed:
Major enhancements

[Enhance] - Added DTO classes for APIs requests & responses
[Enhance] - Changed project status class field type to ENUM
[Refactor] - Made new project API to return created project entity

Htrap19 committed:
Added missing endpoints (/assets, /dependencies, /databaseused, /projectSoftDetails)

Mknan-fms reviewed (approved):
- No need to use `new ArrayList` when returning collection, repository `.findAll()` already returns a list
- When developing an API that has two path variables, consider proceeding the variable with a noun of its entity
- Repetitive calls of the same code blocks, are better to be put in a single utility function and called consecutively. E.g., `.findEntityById` method can wrapped in `private` method
